### PR TITLE
Add PBKDF1 to the legacy provider

### PR DESCRIPTION
**Pull request by jon-oracle**
Added the PKCS#5 key derivation function to the default provider. 

This work precedes some other work to add library context arguments to the PKCS12 APIs.

EDIT: Moved to the legacy provider instead as this algorithm is not recommended for new applications.

### CODE REVIEW

File: providers/implementations/kdfs/pbkdf1.c
Comment (paulidale) [inline providers/implementations/kdfs/pbkdf1.c:2]: 1999-2021 really?

**Review by paulidale — COMMENTED** [review]
Do we really want a very old legacy algorithm added?

I'd like to see some tests added to evp_test too.

**Review by slontis — COMMENTED** [review]
Looks fine to me..
Someone else not from Oracle will need to approve..
NOTE: This KDF will be used within PKCS5_PBE_keyivgen() where equivalent code currently resides.

**Comment by slontis** [comment]
Small compile error is relevant

File: providers/implementations/kdfs/pbkdf1.c
Comment (slontis) [reply providers/implementations/kdfs/pbkdf1.c:2]: This is the date on the p5_crypt.c file where some of this code is extracted from.

**Comment by slontis** [comment]
> Do we really want a very old legacy algorithm added?

This is a KDF so it really belongs in the provider.

**Comment by paulidale** [comment]
Perhaps it could start in the legacy provider :)

**Comment by t8m** [comment]
I agree with Paul that this really belongs to legacy provider.

File: crypto/evp/p5_crpt.c
Comment (slontis) [inline crypto/evp/p5_crpt.c]: you may be able to pass the params into the derive here..

Comment (paulidale) [reply crypto/evp/p5_crpt.c]: It's possible as of last night.

**Comment by slontis** [comment]
Is this still WIP? changes name if it is not and maybe do a p i n g.

File: crypto/evp/p5_crpt.c
Comment (mattcaswell) [inline crypto/evp/p5_crpt.c:98]: Why are these temporary copies necessary? Can't you just directly pass `out` and `out + PKCS5_PBES1_KEY_IV_LENGTH` for the key and iv in `EVP_CipherInit_ex` call below?

File: test/evp_kdf_test.c
Comment (mattcaswell) [inline test/evp_kdf_test.c:366]: Can we have a similar test somewhere for the public API function `PKCS5_PBE_keyivgen` which you've made significant changes to in this PR, but there doesn't seem to be a test for it anywhere.

**Review by mattcaswell — CHANGES_REQUESTED** [review]
This needs a CHANGES entry to at least mention: the introduction of a new EVP_KDF algorithm, the fact that it is in the legacy provider, and that PKCS5_PBE_keyivgen now needs the legacy provider loaded in order to function.

**Comment by jon-oracle** [comment]
> This needs a CHANGES entry to at least mention: the introduction of a new EVP_KDF algorithm, the fact that it is in the legacy provider, and that PKCS5_PBE_keyivgen now needs the legacy provider loaded in order to function.

Rebased and added a note to CHANGES.

**Comment by t8m** [comment]
I think we want to squash the commits before merging as those are mostly work in progress commits during the review.

**Review by paulidale — CHANGES_REQUESTED** [review]
I'd like to see at least a test in the `evp_test`.
I'm not sure why we're persisting with `evp_kdf_test`...

Otherwise, it looks good.

**Review by paulidale — APPROVED** [review]
Perfect, thanks for the extra bit.

**Comment by slontis** [comment]
The error in legacy CI might be valid?

**Comment by slontis** [comment]
Use
````
> perl configdata.pm --dump
````

**Comment by slontis** [comment]
You will notice that the md2 related .txt files for test_evp have this line

````
Availablein = legacy
Digest = MD2
Input = "a"
Output = 32ec01ec4a6dac72c0ab96fb34c0b5d1
````

It needs to be the first line in a test for it to work..

**Comment by jon-oracle** [comment]
> You will notice that the md2 related .txt files for test_evp have this line
> 
> ```
> Availablein = legacy
> Digest = MD2
> Input = "a"
> Output = 32ec01ec4a6dac72c0ab96fb34c0b5d1
> ```
> 
> It needs to be the first line in a test for it to work..

Yes that's already in the test file and as the PBKDF1 algorithm is in legacy it does appear to be loaded. The issue is inside the KDF where ossl_prov_digest_load_from_params() is called and fails to find MD2 despite it being built and available (the evp_test MD2 tests run OK).

Is there any way to list the loaded algorithms for a provider or print some details of the fetch?

**Comment by paulidale** [comment]
`util/wrap.pl apps/openssl list -digest-algorithms -provider legacy` ?

**Comment by t8m** [comment]
You could also try to run just the evp_test with verbose output to see if there are any errors dumped which could give more hints.

**Comment by jon-oracle** [comment]
> You could also try to run just the evp_test with verbose output to see if there are any errors dumped which could give more hints.

The problem is that the library context supplied in the internal call to fetch the digest comes from the provider context and appears to be empty:

```
    OSSL_LIB_CTX *provctx = PROV_LIBCTX_OF(ctx->provctx);

    if (!ossl_prov_digest_load_from_params(&ctx->digest, params, provctx))
        return 0;
```

@slontis has a provisional fix in https://github.com/openssl/openssl/pull/14507

File: providers/implementations/kdfs/pbkdf1.c
Comment (jon-oracle) [reply providers/implementations/kdfs/pbkdf1.c]: Used just after as an argument to ossl_prov_digest_load_from_params(). The name is a bit confusing though - it should be a 'libctx'. I've changed this.

**Comment by mattcaswell** [comment]
Once #14991 is approved and merged this should magically start working.

**Comment by t8m** [comment]
CI failures are relevant.

**Comment by mattcaswell** [comment]
This should fix at least one of the problems:

```` diff
diff --git a/providers/common/build.info b/providers/common/build.info
index ba5df595fe..a14bf09037 100644
--- a/providers/common/build.info
+++ b/providers/common/build.info
@@ -4,5 +4,7 @@ SOURCE[../libcommon.a]=provider_err.c provider_ctx.c
 $FIPSCOMMON=provider_util.c capabilities.c bio_prov.c digest_to_nid.c\
             securitycheck.c provider_seeding.c
 SOURCE[../libdefault.a]=$FIPSCOMMON securitycheck_default.c
-SOURCE[../liblegacy.a]=provider_util.c
+IF[{- !$disabled{module} && !$disabled{shared} -}]
+  SOURCE[../liblegacy.a]=provider_util.c
+ENDIF
 SOURCE[../libfips.a]=$FIPSCOMMON securitycheck_fips.c
````

**Comment by slontis** [comment]
Looks like you have a memory leak in your test.. (Use valgrind locally)

**Comment by slontis** [comment]
test_kdf_pbkdf1 needs to be skipped for the no_legacy case.

**Comment by mattcaswell** [comment]
The memleak seen in the test output is due to #15269.

The other failures are due to a bug in the child libctx code. PR coming up for that. Since the memleak only occurs in an error path which we only hit because of the underlying child libctx bug, fixing that bug is sufficient for this PR.

**Comment by mattcaswell** [comment]
> The other failures are due to a bug in the child libctx code. PR coming up for that.

PR in #15270. With that PR applied to this branch I see the tests that are failing in this PR pass (at least locally).

**Comment by mattcaswell** [comment]
#15270 has been merged. Please rebase again!

**Comment by slontis** [comment]
Jon - when you do a force push like this - can you please add a comment also. e.g I had to rebase- there were no changes made.

**Review by slontis — COMMENTED** [review]
Can you also update doc/man7/OSSL_PROVIDER-legacy.pod

**Comment by jon-oracle** [comment]
> Jon - when you do a force push like this - can you please add a comment also. e.g I had to rebase- there were no changes made. (Or please review the last 2 commits).

No worries.

This was a rebase + make update to update the fips checksums.

**Comment by slontis** [comment]
I am not 100% sure but I think the checksums should not be added anymore.. (There is a different process now - these files will only be updated when a Release in done).

File: doc/man7/OSSL_PROVIDER-legacy.pod
Comment (slontis) [inline doc/man7/OSSL_PROVIDER-legacy.pod:85]: Just ````=item PBKDF1```` is probably sufficient here..

File: providers/fips.checksum
Comment (slontis) [inline providers/fips.checksum:1]: You may be able to remove this file?

**Comment by paulidale** [comment]
Confirming that the checksums aren't required anymore.

File: doc/man7/OSSL_PROVIDER-legacy.pod
Comment (slontis) [reply doc/man7/OSSL_PROVIDER-legacy.pod]: Can you resolve things when you fix them please..

**Comment by jon-oracle** [comment]
Removed checksums and updated doc entry.

File: doc/man3/PKCS5_PBE_keyivgen.pod
Comment (slontis) [inline doc/man3/PKCS5_PBE_keyivgen.pod:86]: Looks good Jon..
One last thing.. Should you mention here the PBKDF1 legacy change maybe?
PKCS#5 PBKDF1 key derivation has been moved from PKCS5_PBE_keyivgen() into
   the legacy crypto provider as an EVP_KDF

Comment (jon-oracle) [reply doc/man3/PKCS5_PBE_keyivgen.pod:86]: Done, thanks.

File: doc/man3/PKCS5_PBE_keyivgen.pod
Comment (slontis) [inline doc/man3/PKCS5_PBE_keyivgen.pod:137]: Sorry maybe the history section?

**Review by slontis — COMMENTED** [review]
LGTM.. (I cant be the first approver due to COI)

File: crypto/evp/p5_crpt.c
Comment (t-j-h) [inline crypto/evp/p5_crpt.c:59]: You have changed the values here from 16 to 8 and also wired in that key and iv length must be the same and must be 8 bytes which isn't what the code had before. Why?

Comment (jon-oracle) [reply crypto/evp/p5_crpt.c:59]: The PBES1 algorithm defined in https://datatracker.ietf.org/doc/html/rfc2898#section-6.1 states:

>       3. Separate the derived key DK into an encryption key K consisting
>          of the first eight octets of DK and an initialization vector IV
>          consisting of the next eight octets:
> 
>                  K   = DK<0..7> ,
>                  IV  = DK<8..15> .

This is the constraint I used to determine the acceptable key/IV lengths.

Comment (t-j-h) [reply crypto/evp/p5_crpt.c:59]: And you are applying this in a way that alters the logic - as these are properties of the input cipher.
You are now wiring them to a fixed value which is not what the code did previously. 

It matters little as to what the RFC states for things like this - we have a routine that can operate using any cipher and you are changing the logic of how it processes the input cipher significantly - and that should not be happening in this sort of code update.

Comment (slontis) [reply crypto/evp/p5_crpt.c:59]: Tim are you aware of an actual use case?  Or is this a theoretical argument? I realize this is a public interface.. (albeit a fairly crappy one)

Comment (t-j-h) [reply crypto/evp/p5_crpt.c:59]: It is a public function that takes and cipher and people do unusual things and wiring this to only work for iv==8 and keylength==8 seems rather inappropriate when the PR isn't about reducing the functionality to just what the RFC states.
We don't claim to limit it and 

If we wanted to do that then it should have been explicitly noted and raised and signed off by the OTC - not just mixed up in the middle of a patch that is meant to just be moving code around.

It really stood out as an input parameter validation change that had nothing to do with the patch.

Looking at the history some asserts got turned into these checks in 2019 a05bf83c7964bb3928b323fe356b9f70f105036d and the original asserts from 2002 are in 54a656ef081f72a740c550ebd8099b40b8b5cde0

There also appears to have been RC4 128-bit usage in the code originally (although not in the more recent code).

Comment (jon-oracle) [reply crypto/evp/p5_crpt.c]: Ok, sure, I can see that this being a public function restricts how much we can clean up the implementation.

I've re-introduced the original logic now.

**Review by t-j-h — APPROVED** [review]

File: crypto/evp/p5_crpt.c
Comment (t-j-h) [reply crypto/evp/p5_crpt.c]: Thanks.

Comment (slontis) [reply crypto/evp/p5_crpt.c]: If I remember rightly I think we did this initially because it was trashing over bits of the buffer that it shouldnt if the size was > 8.. Oh well..

Comment (t-j-h) [reply crypto/evp/p5_crpt.c]: That recollection would be in error. The code is not "trashing over bits of the buffer". If it did then that should be separately fixed. If we wanted to change the behaviour that should be separately fixed. When moving code around it is not the time to be altering behaviours like this as that makes for a challenge in terms of tracking changes.

The code was written in a manner that allowed for non 8-byte cipher algorithm use - there is evidence of that usage in the past - and that isn't covered in the RFC - but that isn't all that unusual with this RFCs which were written after-the-fact and are often simplifications or codifications of implementations that were more rigid that necessarily the practice out there in the wild.

The code has an explicit range check for both keylength and ivlength both of which are safe for the entire range of the check if you read the code. 

Although I see a separate potential issue - if the IV length is 16 bytes then we will use the key as the IV as well which is probably not exactly the expected behaviour. But there is no "trashing" that I can see. If you think there is then there should be a separate issue called out for this and it should be resolved independent of this PR.

Comment (slontis) [reply crypto/evp/p5_crpt.c]: Yep. I didnt quite mean a buffer oveflow :)
md_tmp + (16 - ivl) is just wrong with these constraints..
The keylen constraint is the size of the buffer (64 bytes) - so there would be no space for iv in this case.

**Review by paulidale — DISMISSED** [review]

File: test/recipes/05-test_pbe.t
Comment (paulidale) [inline test/recipes/05-test_pbe.t:26]: Better as: `ok(run(test(["pbetest"])), "running PBE test");` and not using Test::Simple and the $ENV{OPENSSL_MODULES} setting.

Comment (jon-oracle) [reply test/recipes/05-test_pbe.t:26]: Does this include the legacy provider or is there another step needed?

**Review by paulidale — COMMENTED** [review]
Is a test case with a fetched cipher useful?

**Review by paulidale — APPROVED** [review]

File: test/pbetest.c
Comment (paulidale) [inline test/pbetest.c:58]: indentation space is in the wrong place

**Review by paulidale — APPROVED** [review]

File: test/evp_kdf_test.c
Comment (slontis) [inline test/evp_kdf_test.c:321]: This doesnt do a NULL check if the malloc fails.

Comment (jon-oracle) [reply test/evp_kdf_test.c:330]: Fixed this and the other cases.

**Comment by jon-oracle** [comment]
Rebased/squashed commits, removed changes to fips checksums. Added some review updates.

File: test/evp_kdf_test.c
Comment (slontis) [inline test/evp_kdf_test.c:619]: The same issue in in the Scrypt code below..

Comment (jon-oracle) [reply test/evp_kdf_test.c:619]: Scrypt test should be OK as it doesn't allocate params

**Comment by slontis** [comment]
@paulidale - needs reapproval

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
Oops, still need a second reviewer.

**Review by slontis — APPROVED** [review]

**Comment by slontis** [comment]
Note libcrypto.num needs a rebase..

**Comment by jon-oracle** [comment]
No worries, rebased.

**Comment by paulidale** [comment]
And merged finally.  Thanks for the addition.

**Pull request closed without merging**
